[BEAM-1101] Remove inconsistencies in Python PipelineOptions

pabloem opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

pabloem commented:
r: @aaltay
Please let me know your opinion regarding my move of `BIGQUERY_API_SERVICE`, etc. to `GoogleCloudOptions`. I did this to avoid having a circular dependency between `GoogleCloudOptions` and `apiclient.py`. I feel like it makes sense to have these constants within the options - but I'll move them back if necessary.

Result of transaction:
* `dataflow_endpoint` - Using the same default for `options.py` and `apiclient.py`.
* `job_name` - Default is 'beamapp-username-date-microseconds'. Test was added.
* `staging_location` and `temp_location` - `staging_location` was the default of `temp_location`. Now it's the other way around, and the tests reflect that.
* `service_account_name`, `service_account_key_file` - These have been removed.
* `machine_type` alias of `worker_machine_type` has been removed.
* `disk_type` alias of `worker_disk_type` has been removed.
* `disk_source_image` has been removed.
* `no_save_main_session` - has been removed.

I have left removal of `no_pipeline_type_check` for a different change because I had trouble removing it and fixing its tests. I have [filed JIRA-1106](https://issues.apache.org/jira/browse/BEAM-1106) to track this.

aaltay commented:
R: @chamikaramj could you review this PR?

@pabloem For each changed option, could you comment on the status of the same option in the Java SDK.

pabloem commented:
Options in the Java SDK:
* `job_name` - Default in Java is 'BeamApp-username-date-randomnumber'. In Python I've changed it to be 'beamapp-username-date-microseconds'.
* `staging_location` and `temp_location` - In Java, `tempLocation` is the default value for `stagingLocation`. In Python I've changed the behavior to be the same.
* `temp_location` - In Java, there is two parameters: `gcpTempLocation`, and `tempLocation`. The difference is that `gcpTempLocation` has a validation for `'gs://'` in the path. Other runners don't use the `temp_location` parameter, so Ahmet and I feel that it's okay to not add a gcp specific one - for now at least.
* `service_account_name`, `service_account_key_file` - These have been removed from the Java SDK, rather using the GCP technologies supported by the auth team in Google. I've changed to be like that in Python.
* `machine_type` alias of `worker_machine_type` has been removed. In java, the option is `workerMachineType`, it has no aliases.
* `disk_type` alias of `worker_disk_type` has been removed. Same as before.
* `no_save_main_session` -  Is a Python specific parameter.
* `disk_source_image` has been removed from the Java SDK.

* `dataflow_endpoint` - The change of this option is only a bit of a code refactoring.

chamikaramj commented on sdks/python/apache_beam/internal/apiclient.py:
```diff
@@ -165,7 +163,7 @@ def __init__(self, packages, options, environment_version):
           dataflow.Package(
               location='%s/%s' % (
                   self.google_cloud_options.staging_location.replace(
-                      'gs:/', STORAGE_API_SERVICE),
```
From PR description - "staging_location and temp_location - In Java, tempLocation is the default value for stagingLocation. In Python I've changed the behavior to be the same.".

Was this not implemented ? I don't see the change here (apiclient.py:328)

Seems like some other changes mentioned in the PR description are not implemented here as well. Can you pls update the PR description to accurately describe this change ?

chamikaramj replied:
Are you sure following code doesn't have to be modified ? This seems to be doing opposite of what is mentioned in PR description.

     if not self.google_cloud_options.temp_location:
       logging.info('Defaulting to the staging_location as temp_location: %s',
                    self.google_cloud_options.staging_location)
       (self.google_cloud_options
        .temp_location) = self.google_cloud_options.staging_location

Update you mentioned seems to be only be doing an error check not changing default. Did you try running a job without specifying the staging location ?

pabloem replied:
Cham, you're so right. Sorry about that. I've fixed that.

pabloem commented:
Jenkins failure is in the Apex Runner. Python tests pass.

chamikaramj commented:
LGTM

pabloem commented:
@robertwb can you merge, please? : )

pabloem commented:
Relaunching test..

robertwb commented:
"Remove inconsistencies" is somewhat vague. Could you document in the CL body what exactly changed (in particular, what is backwards incompatible).

pabloem commented:
@robertwb I have amended the commit message to contain a full summary of the options that have changed.

aaltay commented on sdks/python/apache_beam/utils/pipeline_options_validator.py:
```diff
@@ -144,11 +145,15 @@ def validate_gcs_path(self, view, arg_name):
+      user_name = getpass.getuser().lower()
+      date_component = datetime.utcnow().strftime('%m%d%H%M%S-%f')
+      app_name = 'beamapp'
+      view.job_name = '{}-{}-{}'.format(app_name, user_name, date_component)
```
Should the default be set in the apiclient at job submission time, instead of validation time?

pabloem replied:
After our chat discussion, I moved the code to a new function in the same module. It's not a significant change, but I can move it to apiclient if you prefer that.
The Python argparse package does not take lambdas for default values.

pabloem replied:
I've moved it to apiclient as you suggested. : )

pabloem commented:
@robertwb this is ready : )
